spyder/plugins/completion/plugin.py

```diff
@@ -110,21 +110,31 @@ def receive_response(self, completion_source, req_id, resp):
 
         if req_id not in self.requests:
             return
-        request_responses = self.requests[req_id]
-        request_responses['sources'][completion_source] = resp
-
-        wait_for = self.WAIT_FOR_SOURCE[request_responses['req_type']]
-        if (completion_source not in wait_for or not resp):
-            # Check if there's any work remaining that may return completions,
-            # since we don't have "good" completions from the current response
-            if any(self._is_client_running(source) and
-                   source not in request_responses['sources']
-                   for source in wait_for):
-                return
-        else:
-            del self.requests[req_id]
 
         with QMutexLocker(self.collection_mutex):
+            request_responses = self.requests[req_id]
```

Comment:
Why the changes inside this QMutexLocker are necessary as part of this PR?

Reply:
I think they are to prevent concurrency problems. Is this right @metalogical?

Reply:
Yes, I noticed some issues when both Kite & LSP were disabled, since fallback completions would not end up being shown. So I just moved more code into the locked segment.

Reply:
Ok, thanks @metalogical for the explanation.